crates/oxc_semantic/src/builder.rs

```diff
@@ -124,11 +126,12 @@ impl<'a> SemanticBuilder<'a> {
             current_reference_flag: ReferenceFlag::empty(),
             current_scope_id,
             current_scope_depth: 0,
-            function_stack: vec![],
-            namespace_stack: vec![],
-            nodes: AstNodes::default(),
+            function_stack: Vec::with_capacity(4),
+            // Most TS modules are not nested, and those that are are usually only 1 level deep.
+            namespace_stack: Vec::with_capacity(2),
+            nodes: AstNodes::with_capacity(source_text.len() >> 4),
```

Comment:
@DonIsaac Just a side note: There's no need to write `>> 4` here. When you multiply or divide by a power of 2, the compiler will convert that to a bitshift operation. So `/ 16` is preferable, as it's easier to read for us humans, and you don't lose any perf.